physics/GFS_rrtmgp_thompsonmp_pre.F90

```diff
@@ -169,7 +169,7 @@ subroutine GFS_rrtmgp_thompsonmp_pre_run(nCol, nLev, nTracers, ncnd, doSWrad, do
     do iLay = 1, nLev
        do iCol = 1, nCol  
           if (ltaerosol .and. qc_mp(iCol,iLay) > 1.e-12 .and. nc_mp(iCol,iLay) < 100.) then
-             nc_mp(iCol,iLay) = make_DropletNumber(qc_mp(iCol,iLay)*rho(iCol,iLay), nwfa(iCol,iLay)) * orho(iCol,iLay)
+             nc_mp(iCol,iLay) = make_DropletNumber(qc_mp(iCol,iLay)*rho(iCol,iLay), nwfa(iCol,iLay)*rho(iCol,iLay)) * orho(iCol,iLay)
```

Comment:
Same comment as before. This becomes a 3rd time for possible differences in code to exist.  I don't see it as necessary.

Is this possibly because timestep#1 itself calls radiation before microphysics?  If so, we should resolve this problem permanently like it can be done in WRF using module_initialize_real.F to ensure no mass can exist without a corresponding number for any MP scheme for any species in 2 moments.  It would be far more general and safe that way.

Reply:
I don't recall the reasoning, but it shouldn't be the timestep, because `mp_thompson_init` runs before any of the `_run` routines. @tanyasmirnova @joeolson42 @hannahcbarnes made this cloud-radiation interaction changes originally. I hope they can comment.

Reply:
To address your other comment about 3rd time. If it isn't necessary for rrtmgp, then it isn't necessary for rrtmg and will disappear in both. Let's hear what the GSL developers have to say.

Reply:
I don't know the history why this call to make_dropletnumber() is in here. 
But when using subgrid scale clouds, the cloud-fields are adjusted just prior to calling this routine in https://github.com/NCAR/ccpp-physics/blob/master/physics/module_SGSCloud_RadPre.F90.

My assumption was that the #-concentration needed to be updated consistently before calling radiation?
Hence why it is the same in RRTMGP as RRTMG.

Reply:
I commit earlier on make_DropletNumber. We have to do it every time step for subgrid clouds.

Reply:
Looking at the two suites `FV3_GSD_v0` and `FV3_GSD_v0_RRTMGP`, and given that it is really the logic in the `sgscloud_radpre` that requires recomputing the effective radii, we could consider moving those calculations out of the radiation-specific pre routines and add them to `sgscloud_radpre`? This way we have only place for these calls in the radiation group not N (=number of radiation schemes).

Reply:
@dustinswales Your assumption is correct.

Reply:
The rrtmg_pre code  prepares effective radii for the calls to  progcld* for all microphysics schemes. Therefore it is logical to call calc_effecRad  for THompson MP in this subroutine. RRTMGP has a separate subroutine for Thompson MP: GFS_rrtmgp_thompsonmp_pre.F90.

Reply:
Hmm, you are right.

Reply:
I strongly urge the group to consider coordinating a single place where we treat creation of water droplet number or ice crystal number for sub-grid-scale clouds in a unified single location, not among one or more variants of radiation physics.  I am already handling explicit/grid-scale clouds within my MP scheme in a uniform way.  We should do likewise for SGS clouds.  And, in so doing, we can also permit potential other SGS cloud schemes to do likewise.  In my view, this should be really simple to do in the file Dustin mentioned (module_SGSCloud_RadPre.F90).  Could you also look over my code mods in PR#567 and note that I have included specific size constraints as parameter constants at top of MP-thompson that could be linked via USE ..., ONLY .... statements?  And, in a related note, my changes to the radiative effective radii also removed the internal size constraints such that after calling the size calculation, the MIN/MAX statements are now quite important.  This permits the actual calculation within the subroutine to go beyond the lower/upper limits in contrast to the old behavior.

Reply:
@gthompsnJCSDA
In the RRTMGP enabled physics we collect all of the MP specific information to a single module (this one in the case of Thompson MP), a simple MP-2-radiation interface. There are other modules that do the same thing for different MP/radiation couplings (e.g GFS_rrtmgp_gfdlmp_pre.F90).

Conceptually, the SGS clouds are not related to this step in the process. The SGS clouds come in from other physical parameterizations (e.g. BL and convection) and the role of the sgsCloud_radpre routine is to couple these other MP to the cloud fields, which needs to be done before we couple the MP to the radiation.

I see no problem with importing min/max's if there needs to be bounding on the effective radii. I do exactly that for the RRTMGP particle sizes.